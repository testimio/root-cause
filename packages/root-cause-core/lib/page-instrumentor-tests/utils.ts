import type { RootCausePage } from '../interfaces';
import { PuppeteerPageHooker } from '../PuppeteerPageHooker';
import { TestContextMemory } from '../TestContextMemory';
import { createDateMocker } from '../utils';

export async function runPageInstrumentorTest<TPage extends RootCausePage>(
  testName: string,
  fullTestFileName: string,
  page: TPage,
  testCode: (wrappedPage: TPage) => Promise<void>
): Promise<SetupPageInstrumentorForTestingReturn<TPage>['hooks']> {
  const { pageInstrumentor, wrappedPage, hooks } = await setupPageInstrumentorForTesting(
    testName,
    fullTestFileName,
    page
  );

  pageInstrumentor.start();

  await testCode(wrappedPage);

  await pageInstrumentor.end({
    success: true,
  });

  return hooks;
}

interface SetupPageInstrumentorForTestingReturn<TPage extends RootCausePage> {
  wrappedPage: TPage;
  pageInstrumentor: PuppeteerPageHooker;
  hooks: {
    beforeAllHook: jest.Mock<any, any>;
    afterAllHook: jest.Mock;
    beforeEachHook: jest.Mock;
    afterEachHook: jest.Mock;
  };
}

export async function setupPageInstrumentorForTesting<TPage extends RootCausePage>(
  testName: string,
  fullTestFileName: string,
  page: TPage
): Promise<SetupPageInstrumentorForTestingReturn<TPage>> {
  const testContext = new TestContextMemory(
    '/tmp',
    testName,
    `full ${testName}`,
    fullTestFileName,
    {
      screenshots: false,
      console: false,
      networkLogs: false,
      jestAssertions: false,
      html: false,
    },
    createDateMocker()
  );

  const pageInstrumentor = new PuppeteerPageHooker(testContext, page);

  const wrappedPage = pageInstrumentor.wrapWithProxy(page);

  const beforeAllHook = jest.fn();
  const afterAllHook = jest.fn();

  const beforeEachHook = jest.fn();
  const afterEachHook = jest.fn();

  pageInstrumentor.registerBeforeAllHook(beforeAllHook);
  pageInstrumentor.registerAfterAllHook(afterAllHook);
  pageInstrumentor.registerBeforeHook(beforeEachHook);
  pageInstrumentor.registerAfterHook(afterEachHook);

  return {
    wrappedPage,
    pageInstrumentor,
    hooks: {
      beforeAllHook,
      afterAllHook,
      beforeEachHook,
      afterEachHook,
    },
  };
}

const objPropToClean = ['testContext', 'proxyContext', 'rootPage', 'instrumentedFunctionResult'];
export function cleanJestMockFunctionCalls(calls: any[]): any[] {
  return calls.map((callArg) => {
    const cloned = { ...callArg };

    for (const prop of objPropToClean) {
      if (prop in cloned) {
        cloned[prop] = 'cleaned';
      }
    }

    return cloned;
  });
}
