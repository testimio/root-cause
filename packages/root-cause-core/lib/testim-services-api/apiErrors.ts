type ApiErrorCode =
  | 'E_UNAUTHENTICATED'
  | 'E_CREDENTIALS_DONT_MATCH'
  | 'E_CREATING_EXECUTION_FAILED'
  | 'E_UPLOAD_REQUEST_FAILED'
  | 'E_FILE_NOT_FOUND';
class TestimApiErrorBase extends Error {
  constructor(public code: ApiErrorCode, message: string) {
    super(message);
  }
}
export class TestimApiError extends TestimApiErrorBase {
  constructor(private originalError: Error, message: string, code: ApiErrorCode) {
    super(code, message);
  }
}
export class TestimNotFoundError extends TestimApiErrorBase {
  constructor(message: string) {
    super('E_FILE_NOT_FOUND', message);
  }
}
