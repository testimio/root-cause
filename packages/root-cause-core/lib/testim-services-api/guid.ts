'use strict';

const a = 'qwertyuiopasdfghjklzxcvbnmQWERTYUIOPASDFGHJKLZXCVBNM1234567890';

export function guid(n = 10) {
  let str = '';
  for (let i = 0; i < n; i++) {
    const index = Math.floor(Math.random() * (a.length - 1));
    str += a[index];
  }
  return str;
}
