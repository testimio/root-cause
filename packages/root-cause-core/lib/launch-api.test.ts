import fs from 'fs-extra';
import path from 'path';
import assert from 'assert';
import { readJsonTestSnapshotFile } from './utils';
import ms from 'ms';
import { launchImpl } from './launch';
// eslint-disable-next-line import/no-extraneous-dependencies
import { ensurePrerequisite, getCurrentTest } from '@testim/root-cause-jest';
import { guid } from './testim-services-api/guid';
import {
  getCleanProcessTicksAndRejectionsStackFramePrettyFormatPlugin,
  getCleanAllPathsPrettyFormatPlugin,
} from '@testim/internal-self-tests-helpers';
import type { TestResultFile } from '@testim/root-cause-types';

describe('Launch api test', () => {
  expect.addSnapshotSerializer(getCleanAllPathsPrettyFormatPlugin(process.cwd()));
  expect.addSnapshotSerializer(getCleanProcessTicksAndRejectionsStackFramePrettyFormatPlugin());
  jest.setTimeout(ms('10 seconds'));

  beforeAll(() => {
    ensurePrerequisite();
  });

  it('Launch api test passing', async () => {
    const testInfo = getCurrentTest();

    const mockedDateConstructor: typeof Date = Object.create(Date);

    let nowCallsCounter = 1;
    mockedDateConstructor.now = function mockedNow() {
      return nowCallsCounter++;
    };

    const localRunId = guid();

    await launchImpl(
      {
        testName: testInfo.fullName,
        noServer: true,
        automationLibrary: 'puppeteer',
        browserOptions: {
          headless: true,
          browserArgs: ['--no-sandbox', '--disable-setuid-sandbox'],
        },
        _runId: localRunId,
      },
      async (page) => {
        await page.goto('http://jsbin.testim.io/ner/1');

        await page.click('#test');
        await page.click('#test');
        await page.click('#test');
        await page.click('#test');
        await page.click('#test');
        await page.click('#test');
        await page.click('#reset');
        await page.click('#test');

        const counter = await page.$('#counter');

        if (!counter) {
          throw new Error('counter not found');
        }

        const textProp = await counter.getProperty('innerText');
        const text = await textProp.jsonValue();

        assert.equal(text, '1');
      },
      mockedDateConstructor
    );

    const runResultsDir = path.resolve(__dirname, '..', '.root-cause', 'runs', localRunId);

    const allResultsInDir = (await fs.readdir(runResultsDir)).sort();

    // const testResultsFixture: TestResultFile = await readJsonTestSnapshotFile(path.resolve(__dirname, 'fixtures', 'launch-api-test-passing', 'results.json'));
    const testResults: TestResultFile = await readJsonTestSnapshotFile(
      path.resolve(runResultsDir, allResultsInDir[0], 'results.json')
    );

    expect(testResults).toMatchSnapshot();

    for (const stepResult of testResults.steps) {
      if (stepResult.screenshot) {
        assert.equal(
          await fs.pathExists(
            path.resolve(runResultsDir, allResultsInDir[0], stepResult.screenshot)
          ),
          true,
          `${stepResult.screenshot} is missing`
        );
      }
    }
  });

  it('Launch api test = failing', async () => {
    const mockedDateConstructor: typeof Date = Object.create(Date);

    let nowCallsCounter = 1;
    mockedDateConstructor.now = function mockedNow() {
      return nowCallsCounter++;
    };

    const testInfo = getCurrentTest();

    const localRunId = guid();

    await assert.rejects(() =>
      launchImpl(
        {
          testName: testInfo.fullName,
          noServer: true,
          automationLibrary: 'puppeteer',
          browserOptions: {
            headless: true,
            browserArgs: ['--no-sandbox', '--disable-setuid-sandbox'],
          },
          _runId: localRunId,
        },
        async (page) => {
          await page.goto('http://jsbin.testim.io/ner/1');

          await page.click('#test');
          await page.click('#test');
          await page.click('#test');
          await page.click('#test');
          await page.click('#test');
          await page.click('#test');
          await page.click('#reset');
          await page.click('#test');

          const counter = await page.$('#counter');

          if (!counter) {
            throw new Error('counter not found');
          }

          const textProp = await counter.getProperty('innerText');
          const text = await textProp.jsonValue();

          assert.equal(text, '2');
        },
        mockedDateConstructor
      )
    );

    const runResultsDir = path.resolve(__dirname, '..', '.root-cause', 'runs', localRunId);

    const allResultsInDir = await fs.readdir(runResultsDir);

    const testResults: TestResultFile = await readJsonTestSnapshotFile(
      path.resolve(runResultsDir, allResultsInDir[0], 'results.json')
    );

    expect(testResults).toMatchSnapshot();

    for (const stepResult of testResults.steps) {
      if (stepResult.screenshot) {
        assert.equal(
          await fs.pathExists(
            path.resolve(runResultsDir, allResultsInDir[0], stepResult.screenshot)
          ),
          true,
          `${stepResult.screenshot} is missing`
        );
      }
    }
  });
});
