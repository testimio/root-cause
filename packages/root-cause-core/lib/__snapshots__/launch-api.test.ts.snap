// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Launch api test Launch api test = failing 1`] = `
Object {
  "metadata": Object {
    "branchInfo": "noise_removed:branchInfo",
    "endedTimestamp": 26,
    "fileName": "noise_removed/packages/root-cause-core/lib/launch-api.test.ts",
    "systemInfo": Object {
      "automationFramework": "puppeteer",
      "browser": "chromium",
      "browserPlatform": "noise_removed:browserPlatform",
      "browserVersion": "HeadlessChrome/85.0.4182.0",
      "modelName": "noise_removed:modelName",
      "modelVersion": "noise_removed:modelVersion",
      "pageViewport": Object {
        "height": 600,
        "width": 800,
      },
      "userAgent": "noise_removed:userAgent",
    },
    "testEndStatus": Object {
      "error": Object {
        "message": "'1' == '2'",
        "name": "AssertionError",
        "stack": "AssertionError [ERR_ASSERTION]: '1' == '2'
    at noise_removed/packages/root-cause-core/lib/launch-api.test.ts:129:24
    at Object.launchImpl (noise_removed/packages/root-cause-core/lib/launch.ts:120:9)
    at waitForActual (assert.js:692:5)
    at Function.rejects (assert.js:801:25)
    at Object.<anonymous> (noise_removed/packages/root-cause-core/lib/launch-api.test.ts:100:9)",
      },
      "success": false,
    },
    "testFullName": "1-Launch api test Launch api test = failing",
    "testName": "Launch api test Launch api test = failing",
    "timestamp": 3,
  },
  "steps": Array [
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 5,
      "fnName": "goto",
      "index": 1,
      "name": "goto \\"http://jsbin.testim.io/ner/1\\"",
      "screenshot": "1.page-goto.jpg",
      "startTimestamp": 4,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "                _runId: localRunId,",
          "            },",
          "            async (page) => {",
          "                await page.goto('http://jsbin.testim.io/ner/1');",
          "",
          "                await page.click('#test');",
          "                await page.click('#test');",
        ],
        "column": 17,
        "fromRowNumber": 106,
        "row": 109,
        "sourceFileRelativePath": "lib/launch-api.test.ts",
        "toRowNumber": 112,
      },
      "text": "http://jsbin.testim.io/ner/1",
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 7,
      "fnName": "click",
      "index": 2,
      "name": "click \\"#test\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "2.page-click.jpg",
      "selector": "#test",
      "startTimestamp": 6,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "            async (page) => {",
          "                await page.goto('http://jsbin.testim.io/ner/1');",
          "",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
        ],
        "column": 17,
        "fromRowNumber": 108,
        "row": 111,
        "sourceFileRelativePath": "lib/launch-api.test.ts",
        "toRowNumber": 114,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 9,
      "fnName": "click",
      "index": 3,
      "name": "click \\"#test\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "3.page-click.jpg",
      "selector": "#test",
      "startTimestamp": 8,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "                await page.goto('http://jsbin.testim.io/ner/1');",
          "",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
        ],
        "column": 17,
        "fromRowNumber": 109,
        "row": 112,
        "sourceFileRelativePath": "lib/launch-api.test.ts",
        "toRowNumber": 115,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 11,
      "fnName": "click",
      "index": 4,
      "name": "click \\"#test\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "4.page-click.jpg",
      "selector": "#test",
      "startTimestamp": 10,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
        ],
        "column": 17,
        "fromRowNumber": 110,
        "row": 113,
        "sourceFileRelativePath": "lib/launch-api.test.ts",
        "toRowNumber": 116,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 13,
      "fnName": "click",
      "index": 5,
      "name": "click \\"#test\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "5.page-click.jpg",
      "selector": "#test",
      "startTimestamp": 12,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#reset');",
        ],
        "column": 17,
        "fromRowNumber": 111,
        "row": 114,
        "sourceFileRelativePath": "lib/launch-api.test.ts",
        "toRowNumber": 117,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 15,
      "fnName": "click",
      "index": 6,
      "name": "click \\"#test\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "6.page-click.jpg",
      "selector": "#test",
      "startTimestamp": 14,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#reset');",
          "                await page.click('#test');",
        ],
        "column": 17,
        "fromRowNumber": 112,
        "row": 115,
        "sourceFileRelativePath": "lib/launch-api.test.ts",
        "toRowNumber": 118,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 17,
      "fnName": "click",
      "index": 7,
      "name": "click \\"#test\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "7.page-click.jpg",
      "selector": "#test",
      "startTimestamp": 16,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#reset');",
          "                await page.click('#test');",
          "",
        ],
        "column": 17,
        "fromRowNumber": 113,
        "row": 116,
        "sourceFileRelativePath": "lib/launch-api.test.ts",
        "toRowNumber": 119,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 19,
      "fnName": "click",
      "index": 8,
      "name": "click \\"#reset\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "8.page-click.jpg",
      "selector": "#reset",
      "startTimestamp": 18,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#reset');",
          "                await page.click('#test');",
          "",
          "                const counter = await page.$('#counter');",
        ],
        "column": 17,
        "fromRowNumber": 114,
        "row": 117,
        "sourceFileRelativePath": "lib/launch-api.test.ts",
        "toRowNumber": 120,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 21,
      "fnName": "click",
      "index": 9,
      "name": "click \\"#test\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "9.page-click.jpg",
      "selector": "#test",
      "startTimestamp": 20,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#reset');",
          "                await page.click('#test');",
          "",
          "                const counter = await page.$('#counter');",
          "",
        ],
        "column": 17,
        "fromRowNumber": 115,
        "row": 118,
        "sourceFileRelativePath": "lib/launch-api.test.ts",
        "toRowNumber": 121,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 23,
      "fnName": "getProperty",
      "index": 10,
      "name": "getProperty",
      "screenshot": "10.elementhandle-getProperty.jpg",
      "startTimestamp": 22,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "                    throw new Error('counter not found');",
          "                }",
          "",
          "                const textProp = await counter.getProperty('innerText');",
          "                const text = await textProp.jsonValue();",
          "",
          "                assert.equal(text, '2');",
        ],
        "column": 34,
        "fromRowNumber": 123,
        "row": 126,
        "sourceFileRelativePath": "lib/launch-api.test.ts",
        "toRowNumber": 129,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "endTimestamp": 25,
      "fnName": "Assertion Error",
      "index": 11,
      "name": "Assertion Error \\"'1' == '2'\\"",
      "rect": Object {
        "error": "possible noise removed",
      },
      "screenshot": null,
      "startTimestamp": 24,
      "stepError": Object {
        "message": "'1' == '2'",
        "name": "AssertionError",
        "stack": "AssertionError [ERR_ASSERTION]: '1' == '2'
    at noise_removed/packages/root-cause-core/lib/launch-api.test.ts:129:24
    at Object.launchImpl (noise_removed/packages/root-cause-core/lib/launch.ts:120:9)
    at waitForActual (assert.js:692:5)
    at Function.rejects (assert.js:801:25)
    at Object.<anonymous> (noise_removed/packages/root-cause-core/lib/launch-api.test.ts:100:9)",
      },
      "text": "'1' == '2'",
    },
  ],
}
`;

exports[`Launch api test Launch api test passing 1`] = `
Object {
  "metadata": Object {
    "branchInfo": "noise_removed:branchInfo",
    "endedTimestamp": 24,
    "fileName": "noise_removed/packages/root-cause-core/lib/launch-api.test.ts",
    "systemInfo": Object {
      "automationFramework": "puppeteer",
      "browser": "chromium",
      "browserPlatform": "noise_removed:browserPlatform",
      "browserVersion": "HeadlessChrome/85.0.4182.0",
      "modelName": "noise_removed:modelName",
      "modelVersion": "noise_removed:modelVersion",
      "pageViewport": Object {
        "height": 600,
        "width": 800,
      },
      "userAgent": "noise_removed:userAgent",
    },
    "testEndStatus": Object {
      "success": true,
    },
    "testFullName": "1-Launch api test Launch api test passing",
    "testName": "Launch api test Launch api test passing",
    "timestamp": 3,
  },
  "steps": Array [
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 5,
      "fnName": "goto",
      "index": 1,
      "name": "goto \\"http://jsbin.testim.io/ner/1\\"",
      "screenshot": "1.page-goto.jpg",
      "startTimestamp": 4,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "                _runId: localRunId,",
          "            },",
          "            async (page) => {",
          "                await page.goto('http://jsbin.testim.io/ner/1');",
          "",
          "                await page.click('#test');",
          "                await page.click('#test');",
        ],
        "column": 17,
        "fromRowNumber": 43,
        "row": 46,
        "sourceFileRelativePath": "lib/launch-api.test.ts",
        "toRowNumber": 49,
      },
      "text": "http://jsbin.testim.io/ner/1",
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 7,
      "fnName": "click",
      "index": 2,
      "name": "click \\"#test\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "2.page-click.jpg",
      "selector": "#test",
      "startTimestamp": 6,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "            async (page) => {",
          "                await page.goto('http://jsbin.testim.io/ner/1');",
          "",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
        ],
        "column": 17,
        "fromRowNumber": 45,
        "row": 48,
        "sourceFileRelativePath": "lib/launch-api.test.ts",
        "toRowNumber": 51,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 9,
      "fnName": "click",
      "index": 3,
      "name": "click \\"#test\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "3.page-click.jpg",
      "selector": "#test",
      "startTimestamp": 8,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "                await page.goto('http://jsbin.testim.io/ner/1');",
          "",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
        ],
        "column": 17,
        "fromRowNumber": 46,
        "row": 49,
        "sourceFileRelativePath": "lib/launch-api.test.ts",
        "toRowNumber": 52,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 11,
      "fnName": "click",
      "index": 4,
      "name": "click \\"#test\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "4.page-click.jpg",
      "selector": "#test",
      "startTimestamp": 10,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
        ],
        "column": 17,
        "fromRowNumber": 47,
        "row": 50,
        "sourceFileRelativePath": "lib/launch-api.test.ts",
        "toRowNumber": 53,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 13,
      "fnName": "click",
      "index": 5,
      "name": "click \\"#test\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "5.page-click.jpg",
      "selector": "#test",
      "startTimestamp": 12,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#reset');",
        ],
        "column": 17,
        "fromRowNumber": 48,
        "row": 51,
        "sourceFileRelativePath": "lib/launch-api.test.ts",
        "toRowNumber": 54,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 15,
      "fnName": "click",
      "index": 6,
      "name": "click \\"#test\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "6.page-click.jpg",
      "selector": "#test",
      "startTimestamp": 14,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#reset');",
          "                await page.click('#test');",
        ],
        "column": 17,
        "fromRowNumber": 49,
        "row": 52,
        "sourceFileRelativePath": "lib/launch-api.test.ts",
        "toRowNumber": 55,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 17,
      "fnName": "click",
      "index": 7,
      "name": "click \\"#test\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "7.page-click.jpg",
      "selector": "#test",
      "startTimestamp": 16,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#reset');",
          "                await page.click('#test');",
          "",
        ],
        "column": 17,
        "fromRowNumber": 50,
        "row": 53,
        "sourceFileRelativePath": "lib/launch-api.test.ts",
        "toRowNumber": 56,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 19,
      "fnName": "click",
      "index": 8,
      "name": "click \\"#reset\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "8.page-click.jpg",
      "selector": "#reset",
      "startTimestamp": 18,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#reset');",
          "                await page.click('#test');",
          "",
          "                const counter = await page.$('#counter');",
        ],
        "column": 17,
        "fromRowNumber": 51,
        "row": 54,
        "sourceFileRelativePath": "lib/launch-api.test.ts",
        "toRowNumber": 57,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 21,
      "fnName": "click",
      "index": 9,
      "name": "click \\"#test\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "9.page-click.jpg",
      "selector": "#test",
      "startTimestamp": 20,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "                await page.click('#test');",
          "                await page.click('#test');",
          "                await page.click('#reset');",
          "                await page.click('#test');",
          "",
          "                const counter = await page.$('#counter');",
          "",
        ],
        "column": 17,
        "fromRowNumber": 52,
        "row": 55,
        "sourceFileRelativePath": "lib/launch-api.test.ts",
        "toRowNumber": 58,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 23,
      "fnName": "getProperty",
      "index": 10,
      "name": "getProperty",
      "screenshot": "10.elementhandle-getProperty.jpg",
      "startTimestamp": 22,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "                    throw new Error('counter not found');",
          "                }",
          "",
          "                const textProp = await counter.getProperty('innerText');",
          "                const text = await textProp.jsonValue();",
          "",
          "                assert.equal(text, '1');",
        ],
        "column": 34,
        "fromRowNumber": 60,
        "row": 63,
        "sourceFileRelativePath": "lib/launch-api.test.ts",
        "toRowNumber": 66,
      },
      "unhandledExceptions": Array [],
    },
  ],
}
`;
